### ISSUE TITLE
HDR control for Razer Kiyo Pro

### ISSUE THREAD

**Issue by juliussohn** [Itaybre/CameraController#33]
Hey unfortunately the Razer Synapse Software to configure the webcam only works on windows. Camera Controller already covers most controls, but one key feature of the Razer Kiyo Pro cam is HDR, which would be awesome to have in camera controller

**Comment by Itaybre** [comment]
I would need to see the data sent to the device to see if I can replicate it.

Do you mind extracting it with this software: https://freeusbanalyzer.com

**Comment by Itaybre** [comment]
I just found that HDR is part of the UVC protocol, I will take a look.
But since I cannot test it, I can't promise anything.
I will need the output from USB Prober, see this small guide: https://github.com/Itaybre/CameraController/issues/25#issuecomment-953162336

**Comment by juliussohn** [comment]
Does this help?

<img width="798" alt="CleanShot 2022-01-11 at 14 01 51@2x" src="https://user-images.githubusercontent.com/4502656/148947209-6a83eda4-d3b9-40b8-87e6-2e914613d5ce.png">

Also here are the settings from the Windows Software
<img width="1659" alt="CleanShot 2022-01-11 at 13 42 58@2x" src="https://user-images.githubusercontent.com/4502656/148947236-0ee15983-f279-46e6-80d1-9a8caace0660.png">

**Comment by Itaybre** [comment]
> Does this help?
> 
> 
> 
> <img width="798" alt="CleanShot 2022-01-11 at 14 01 51@2x" src="https://user-images.githubusercontent.com/4502656/148947209-6a83eda4-d3b9-40b8-87e6-2e914613d5ce.png">
> 
> 
> 
> Also here are the settings from the Windows Software
> 
> <img width="1659" alt="CleanShot 2022-01-11 at 13 42 58@2x" src="https://user-images.githubusercontent.com/4502656/148947236-0ee15983-f279-46e6-80d1-9a8caace0660.png">
> 
> 
> 
> 

Yes, that work, but can you expand the "Controls Supported" in the VDC Input Terminal section?

**Comment by juliussohn** [comment]
<img width="780" alt="CleanShot 2022-01-11 at 15 14 22@2x" src="https://user-images.githubusercontent.com/4502656/148958835-9daeda62-7847-4983-a638-faf3477676cc.png">

**Comment by Itaybre** [comment]
I may have found some insight about how to control HDR and also FOV.

I will need help testing that, I couldn't find the camera at a reasonable price locally

**Comment by juliussohn** [comment]
Sure, happy to test if you have a build!

**Comment by Itaybre** [comment]
@juliussohn You should see a new section inside Advanced, for nor I just implemented setting HDR, please tell me if it works correctly

[CameraControllerRazer.zip](https://github.com/Itaybre/CameraController/files/7899597/CameraControllerRazer.zip)

**Comment by juliussohn** [comment]
I cant see an HDR setting. Am I missing something?
![CleanShot 2022-01-21 at 10 30 26@2x](https://user-images.githubusercontent.com/4502656/150502522-1bb75678-80fc-4295-b9da-5f07e342a27d.png)

**Comment by Itaybre** [comment]
@juliussohn Seems like I used an incorrect value for productId and vendorId (to recognize the camera).
Do you mind sharing the values from USB Prober?
They should be at the top when selecting the camera

**Comment by juliussohn** [comment]
Still nothing showing up

**Comment by Itaybre** [comment]
@juliussohn found the issue, it should appear now
[CameraController.zip](https://github.com/Itaybre/CameraController/files/7917382/CameraController.zip)

**Comment by juliussohn** [comment]
@Itaybre  The toggle is there, but it doesn't seem to change the HDR setting

**Comment by Corbot5000** [comment]
I can confirm that turning HDR on and off doesn't change anything. It's worth noting that the Kiyo Pro only supports HDR at 30 FPS - could this be why it isn't working correctly? Is there any way to manually adjust the cameras to 30 FPS within CameraConttoller?

Happy to do what I can to test this functionality.

**Comment by Itaybre** [comment]
@Corbot5000 I didn't know about the 30FPS rule.
I would need recordings made with [Wireshark](https://www.wireshark.org), toggling the FPS up and down.
And probably more recordings of enabling and disabling HDR.

I don't have such camera to test it myself, and it isn't even sold locally

**Comment by Corbot5000** [comment]
@Itaybre I have wireshark installed and the window's camera app lets me toggle between "hdr pro" and "hdr off" when I take photos. If you could give me a little more guidance on how to analyze the Kiyo pro in Wireshark, I'd be happy to do it. I can be reached on Discord at Corbot8000#8015

**Comment by Corbot5000** [comment]
@Itaybre attached are two wireshark logs -
1) KiyoProHDRProOn.pcapng - HDR Pro already turned on, a few photos taken, video recorded
2) KiyoProHDROff.pcapng - Turned HDR Pro off, photos taken, video recorded, HDR Pro turned back on

https://drive.google.com/file/d/1b6Jehiyer1171aHhA1ePDN-09gIP-Iyk/view?usp=sharing

**Comment by Itaybre** [comment]
@Corbot5000 can you do the recording as short as possible? Like just pressing the toggle in the app and wait a couple seconds doing nothing
seeing too many rows doesn't help me find the correct ones

**Comment by Corbot5000** [comment]
@Itaybre Here's the smallest recording I could do of HDR Pro being turned on and off.

https://drive.google.com/file/d/1jLbzS2a_tifFFbSZsh2uj9qDH58_bwqf/view?usp=sharing

**Comment by carlosnasillo** [comment]
Have a Razer Kiyo Pro myself (on Mac M1), and happy to help debug or provide test data as required.

**Comment by lupus2k** [comment]
Razer Kiyo Pro and Mac M1 Max user here, let me know how can I help testing.

**Comment by authorfunction** [comment]
Hi, ordered a Razer Kiyo Pro recently and will be happy to help with testing as soon as it arrives.

**Comment by soyersoyer** [comment]
You can find some working code to set the Kiyo Pro custom controls: https://github.com/soyersoyer/kiyoproctrls

**Comment by authorfunction** [comment]
I tried the code @soyersoyer provided, but there's no /dev/video0 or any other obvious video devices in the /dev directory. The camera is on and working (I tried with it off as well). I am on Monterey (Apple Silicon).

**Comment by jaketame** [comment]
I have a Kiyo Pro now and would be happy to test

**Comment by mattheworiordan** [comment]
I would be happy to test too 👍

**Comment by mk3rnb8** [comment]
Hi All, is there any progress? I'd love to use HDR with my MBP.

cheers,
M.

**Comment by chansearrington** [comment]
Joining in on this thread. I just got two Kiyo Pro Ultras and love them. But it's hard to adjust anything on my Mac. Thank god for this app and eCamm Live

**Comment by retpolanne** [comment]
Hey @Itaybre, would you mind sharing the HDR work on a branch so we can help? As @soyersoyer mentioned, there's this repo with some stuff for HDR and FoV that we can try to implement here

https://github.com/soyersoyer/kiyoproctrls/blob/main/kiyoproctrls.py

**Comment by Itaybre** [comment]
> Hey @Itaybre, would you mind sharing the HDR work on a branch so we can help? As @soyersoyer mentioned, there's this repo with some stuff for HDR and FoV that we can try to implement here
> 
> 
> 
> https://github.com/soyersoyer/kiyoproctrls/blob/main/kiyoproctrls.py

I changed computers several times since this, so I might have lost the code.
But seeing there is a project with this working, it should be much easier to implement. Will try making a build this week

**Comment by alexbochel** [comment]
I would love to second adding this feature. I am a bit rusty with xcode development but I will see if I can find a solution using the linux versions of this app.

**Comment by padurean** [comment]
Are there any news about this?
